src/util/_managing_headers.py

```diff
@@ -93,3 +88,39 @@ def form_additional_headers(ctx: contextvars.Context = None) -> Dict[str, str]:
         'X-Top-Level-Method': ctx.get(top_level_method_var),
         'X-Low-Level-Method': ctx.get(low_level_method_var),
     }
+
+
+def run_in_current_context(func, *args, **kwargs):
+    """Runs the function using context state from the moment of calling run_in_current_context function.
+
+    Unlike Context.run supports generators, async generators and functions that return an awaitable (e.g. coroutines).
+    """
+
+    result = func(*args, **kwargs)
+    if inspect.isawaitable(result):
```

Comment:
I thought contextvars work with coroutines by default ([asyncio support](https://docs.python.org/3/library/contextvars.html)), why do you think it is not true or I don't understand smth?

Reply:
Yes, but "supported by default" simply means that the context is copied on the asyncio.Task creation (but not on coroutine creation). Here is a snippet that proves it:

```
import asyncio
from contextvars import ContextVar

my_var = ContextVar('my_var')

async def a():
    assert my_var.get()


async def main():
    my_var.set(True)
    a_coroutine = a()
    my_var.set(False)
    try:
        await a_coroutine
    except AssertionError:
        pass
    else:
        assert 'I am wrong'

    my_var.set(True)
    a_coroutine = asyncio.create_task(a())
    my_var.set(False)
    await a_coroutine


asyncio.run(main())
```